### PR TITLE
stats: rename TestSymbolTable to be inside Stats::TestUtil, and add help text to a common assert pointing to that as a remedy.

### PR DESCRIPTION
**Pull request by jmarantz**
Commit Message: An earlier refactor required moving `TestSymbolTable` from the mocking infrastructure into `test/common/stats/stat_test_utility.h`, but leaving it outside the `TestUtil` namespace to avoid making a large PR even larger. This makes the namespace consistent again.

More importantly, a common reason to use `TestSymbolTable` is an assertion that arises in tests where names from two different symbol tables get joined(). We now annotate this assertion with a pointer to a section in stats.md where we explain why that assertion occurs and how to remedy it.
Additional Description:
Risk Level: low
Testing: //test/..., plus I manually tested by temporarily changing the stats_store in logical_dns_cluster_test.cc to an Isolated StoreImpl, to watch the assert fire and the message that comes out.
Docs Changes: stats.md updated here.
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: source/docs/stats.md [hunk 1]
```diff
@@ -262,3 +262,30 @@ Developers trying to can iterate through changes in these tests locally with:
       test/integration:stats_integration_test
 ```
 
+## Debugging Symbol Table Assertions
+
+If you are visiting this section because you saw a message like:
+
+```bash
+[2020-12-11 18:30:54.059][16][critical][assert] [source/common/stats/symbol_table_impl.cc:251] assert failure: decode_search != decode_map_.end(). Details: Please see https://github.com/envoyproxy/envoy/blob/master/source/docs/stats.md#debugging-symbol-table-asserts for hints on how to fix
```
Comment (pgenera) [inline source/docs/stats.md:270]: Removing the timestamp will require a bit less horizontal scrolling on the rendered page.

File: source/docs/stats.md [hunk 2]
```diff
@@ -262,3 +262,30 @@ Developers trying to can iterate through changes in these tests locally with:
...
+
+```
+then you have come to the right place.
+
+In production, there is generally one `SymbolTable` per process, except in the 3
```
Comment (pgenera) [inline source/docs/stats.md:275]: is there one `SymbolTable` per process or per thread (or some other abstraction I'm unaware of)?

Comment (jmarantz) [reply source/docs/stats.md:275]: One per process.

File: source/docs/stats.md (hunk 1 above)
Comment (jmarantz) [reply source/docs/stats.md:270]: Good catch; actually I'll go ahead and add some newlines for doc purposes.

File: source/common/stats/symbol_table_impl.cc [hunk 3]
```diff
@@ -243,10 +243,16 @@ void SymbolTableImpl::incRefCount(const StatName& stat_name) {
...
     auto decode_search = decode_map_.find(symbol);
-    ASSERT(decode_search != decode_map_.end());
 
+    ASSERT(decode_search != decode_map_.end(),
+           "Please see "
+           "https://github.com/envoyproxy/envoy/blob/master/source/docs/stats.md#"
+           "debugging-symbol-table-asserts");
     auto encode_search = encode_map_.find(decode_search->second->toStringView());
-    ASSERT(encode_search != encode_map_.end());
+    ASSERT(encode_search != encode_map_.end(),
```
Comment (pgenera) [inline source/common/stats/symbol_table_impl.cc:252]: A lookup failure in the encode map but not the decode would be... doubly confusing, right? I presume these maps are kept in sync elsewhere.

This is fine, just that the second assert seems like it'd point to something more nefarious than the first.

**Review by pgenera — APPROVED** [review]

File: source/common/stats/symbol_table_impl.cc (hunk 3 above)
Comment (jmarantz) [reply source/common/stats/symbol_table_impl.cc:252]: yeah these maps are kept in sync under lock, so I don't expect that could happen.

**Review by snowp — APPROVED** [review]
LGTM, thanks!

**Pull request merged**